### PR TITLE
#102 FIX

### PR DESCRIPTION
**Pull request by TikhomirovSergey**
## Change list
- MissingParameterException was removed
- some code improvements in MultiTouchAction. Now it will throw `IllegalArgumentException` when no action is defined.
 
#102

## Types of changes

- [ ] No changes in production code.
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
+ 💯 for picking up issues that have been waiting for three years in the backlog 👍

File: src/main/java/io/appium/java_client/MultiTouchAction.java [hunk 1]
```diff
@@ -62,16 +65,14 @@ public MultiTouchAction add(TouchAction action) {
...
-        if (size > 1) {
+        List<TouchAction> touchActions = actions.build();
+        checkArgument(touchActions.size() > 0,
+                "MultiTouch action must have at least one TouchAction added before it can be performed");
+        if (touchActions.size() > 1) {
             performsTouchActions.performMultiTouchAction(this);
-        } else if (size == 1) {
+        } else {
             //android doesn't like having multi-touch actions with only a single TouchAction...
             performsTouchActions.performTouchAction(actions.build().get(0));
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/MultiTouchAction.java]: perhaps `touchActions.get(0)` ?

Comment (TikhomirovSergey) [reply src/main/java/io/appium/java_client/MultiTouchAction.java]: yes. Will improve soon

File: src/main/java/io/appium/java_client/MultiTouchAction.java [hunk 2]
```diff
@@ -62,16 +65,14 @@ public MultiTouchAction add(TouchAction action) {
      * Perform the multi-touch action on the mobile performsTouchActions.
      */
     public MultiTouchAction perform() {
-        int size = actions.build().size();
-        if (size > 1) {
+        List<TouchAction> touchActions = actions.build();
+        checkArgument(touchActions.size() > 0,
+                "MultiTouch action must have at least one TouchAction added before it can be performed");
+        if (touchActions.size() > 1) {
             performsTouchActions.performMultiTouchAction(this);
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/MultiTouchAction.java]: you can `returns this` immediately here, so then `else` block is not needed

File: src/main/java/io/appium/java_client/MultiTouchAction.java [hunk 3]
```diff
@@ -80,9 +81,8 @@ public MultiTouchAction perform() {
         ImmutableList.Builder<Object> listOfActionChains = ImmutableList.builder();
         ImmutableList<TouchAction> touchActions = actions.build();
 
-        touchActions.forEach(action -> {
-            listOfActionChains.add(action.getParameters().get("actions"));
-        });
+        touchActions.forEach(action ->
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/MultiTouchAction.java]: why not to replace it with something like `ImmutableList.Builder<Object> listOfActionChains = ImmutableList.copyOf(touchActions.stream().map(x -> x.getParameters().get("actions")));` ?

File: src/main/java/io/appium/java_client/MultiTouchAction.java (hunk 2 above)
Comment (TikhomirovSergey) [reply src/main/java/io/appium/java_client/MultiTouchAction.java]: @mykola-mokhnach 
I can't understand how to do it
```java
performsTouchActions.performMultiTouchAction(this);
```
is `void`

```java
performsTouchActions.performTouchAction(touchActions.get(0));
```
returns `TouchAction`

**Comment by TikhomirovSergey** [comment]
@mykola-mokhnach There is one more change.
Yes. There is some legacy (from Java 7 times) code which still can be here. Sometimes it is hard to see it because of force of habit :)
@SrinivasanTarget your rewiev is is also valuable.

File: src/main/java/io/appium/java_client/MultiTouchAction.java (hunk 2 above)
Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/MultiTouchAction.java]:
```java
        if (touchActions.size() > 1) {
            performsTouchActions.performMultiTouchAction(this);
            resturn this;
        }
        //android doesn't like having multi-touch actions with only a single TouchAction...
        performsTouchActions.performTouchAction(touchActions.get(0));
        return this;
```

**Pull request merged**